vernac/metasyntax.ml

```diff
@@ -1473,6 +1475,55 @@ let check_locality_compatibility local custom i_typs =
                     strbrk " which is local."))
       (List.uniquize allcustoms)
 
+let longest_common_prefix_level ntn =
+  Notgram_ops.longest_common_prefix ntn
+  |> Option.map (fun (ntn, sz) ->
+         let level, levels = level_firstn sz (Notation.level_of_notation ntn) in
+         ntn, level.notation_level, levels)
+
+let default_prefix_level ntn =
+  let with_prefix prefix level =
+    Flags.if_verbose Feedback.msg_info
+      (strbrk "Setting notation at level " ++ int level ++ spc ()
+       ++ str "to match previous notation with longest common prefix:"
+       ++ spc () ++ str "\"" ++ str (snd prefix) ++ str "\".");
+    level in
+  function Some n -> Some n | None ->
+    longest_common_prefix_level ntn
+    |> Option.map (fun (prefix, level, _) -> with_prefix prefix level)
+
+let default_prefix_level_subentries ntn symbols etyps =
+  let with_prefix prefix from_level levels =
+    let default_entry etyps (x, l) =
+      let l' = match l with
+        | LevelLt n when Int.equal n from_level -> NextLevel
+        | LevelLe n | LevelLt n -> NumLevel n
+        | LevelSome -> DefaultLevel in
+      let e = List.assoc_opt x etyps
+        |> Option.default (ETConstr (fst ntn, None, DefaultLevel)) in
+      match l', e with
+      | (NumLevel _ | NextLevel), ETConstr (n, b, DefaultLevel) ->
+         Flags.if_verbose Feedback.msg_info
+           (strbrk "Setting " ++ Id.print x ++ str " "
+            ++ pr_arg_level from_level (l, e) ++ spc ()
+            ++ str "to match previous notation with longest common prefix:"
+            ++ spc () ++ str "\"" ++ str (snd prefix) ++ str "\".");
+         (x, ETConstr (n, b, l')) :: List.remove_assoc x etyps
+      | _ -> etyps in
+    let levels =
+      let rec aux levs symbs = match levs, symbs with
+        | [], _ | _, [] | _, SProdList _ :: _ -> []
+        (* not handling recursive notations *)
+        | _, (Terminal _ | Break _) :: symbs -> aux levs symbs
+        | l :: levs, NonTerminal x :: symbs -> (x, l) :: aux levs symbs in
+      match levels, symbols with
+      (* don't mess up with level of left border terminal *)
+      | _ :: levs, NonTerminal _ :: symbs | levs, symbs -> aux levs symbs in
+    List.fold_left default_entry etyps levels in
+  match longest_common_prefix_level ntn with
```

Comment:
Maybe the two calls to `longest_common_prefix_level` can be factorized?

Reply:
I thought about it before deciding it wouldn't be worth it, but why not.